src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +72,215 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
+    Log::Assert(input.n_cols > 0, "Empty input. Batch size must be \
+        greater than zero.");
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+    Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+        by feature maps.");
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; ++i)
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
+      // Normalize the input.
+      output = input.each_col() - mean;
+      inputMean = output;
+      output.each_col() /= arma::sqrt(variance + eps);
+
+      // Use Welford method to compute the sample variance and mean.
+      for (size_t i = 0; i < input.n_cols; i++)
+      {
+        count += 1;
+
+        OutputDataType diff = input.col(i) - runningMean;
+        if (average)
+        {
+          runningMean = runningMean + diff / count;
+          runningVariance += diff % (input.col(i) - runningMean);
+        }
+        else
+        {
+          runningMean = (1 - momentum) * runningMean + momentum * mean;
+          runningVariance = (1 - momentum) * runningVariance +
+              momentum * variance;
+        }
+      }
+
+      // Reused in the backward and gradient step.
+      normalized = output;
 
-      OutputDataType diff = input.col(i) - runningMean;
-      runningMean = runningMean + diff / count;
-      runningVariance += diff % (input.col(i) - runningMean);
+      // Scale and shift the output.
+      output.each_col() %= gamma;
+      output.each_col() += beta;
+      return;
     }
 
-    // Reused in the backward and gradient step.
-    normalized = output;
+    // Input corresponds to output from convolution layer.
+    // Use a cube for simplicity.
+    size_t batchSize = input.n_cols;
+    arma::cube inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(),
+        input.n_rows / size, size, batchSize, false, false);
+
+    // Initialize output to same size and values for convience.
+    output.set_size(arma::size(input));
+    arma::cube outputTemp(const_cast<arma::Mat<eT>&>(output).memptr(),
+        input.n_rows / size, size, input.n_cols, false, false);
+    outputTemp = inputTemp;
+
+    // Calculate mean and variance over all channels.
+    mean = arma::mean(arma::mean(inputTemp, 2), 0);
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        input.n_rows / size, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, input.n_rows / size, 1);
+
+    // Used in backward propogation.
+    inputMean.clear();
```

Comment:
do we need to clear this everytime?

Reply:
I think just set_size would work fine. I'll remove it. Thanks.